utilities/write_batch_with_index/write_batch_with_index_internal.cc

```diff
@@ -802,4 +802,88 @@ WBWIIteratorImpl::Result WriteBatchWithIndexInternal::GetFromBatch(
   return result;
 }
 
+WBWIIteratorImpl::Result WriteBatchWithIndexInternal::GetEntityFromBatch(
```

Comment:
I see that what we are doing here is nearly identical to `WriteBatchWithIndexInternal::GetFromBatch`

I'm wondering if we could build something like `GetFromBatchCommon` which takes both `value` and `columns` pointer, and set them as necessary.

Reply:
There are certainly some shared snippets but I wouldn't say the two are nearly nearly identical. The most important difference is in the handling of the "found" case when there are no merge operands in the context: in this case, `GetFromBatch` copies the value (for plain values) or the value of the default column (for entities), while `GetEntityFromBatch` calls one of `PinnableWideColumns::{SetPlainValue,SetWideColumnValue}`. Of course, there are ways to deduplicate the shared parts but they might have drawbacks in terms of readability. For instance, if we want to have an efficient implementation of `GetFromBatchCommon` that doesn't do redundant checks, we would have to make it a template that takes function objects that perform these operations - I feel we might end up with more code and more complexity that way.

Reply:
Taking both `value` and `columns` pointers and setting them accordingly while sharing rest of other parts are what we are doing in `db_impl.cc`, (e.g. both `Get()` and `GetEntity()` calling `GetImpl()` internally) and I found that the benefit is bigger than the readability that we slight lose by doing so.

For instance, if we later decide to return a different Status than `Corrupt` or change the error message when `kError` is returned by `iter->FindLatestUpdate()`, we will have to update in two different places.

Reply:
> Taking both `value` and `columns` pointers and setting them accordingly while sharing rest of other parts are what we are doing in `db_impl.cc`, (e.g. both `Get()` and `GetEntity()` calling `GetImpl()` internally) and I found that the benefit is bigger than the readability that we slight lose by doing so.

That solution, while it works, is suboptimal in the sense that it does unnecessary work: at the call site, we know whether the user passed in a `PinnableSlice` (`Get`) or `PinnableWideColumns` (`GetEntity`) but then we lose that information when we call `GetImpl` and have to recover it using runtime checks, which is not ideal. Templates can be used to avoid this but they introduce a bit of complexity. (The `DBImpl` code could technically be also refactored this way but it would be a pretty big undertaking; the `GetImpl` logic predates wide columns and actually goes back a long way.)

Reply:
Got it. Right. Templates could be ideal here, but like you said earlier, it could hurt readability (Honestly I won't be too sure until we actually do it. We will see). Anyway, we can look into this option in a separate PR later.